system/install-service.sh.in

```diff
@@ -0,0 +1,671 @@
+
+  # if /etc/runit/1 exists and is executable, then it's runit
+  [ -x /etc/runit/1 ] && return 0
+
+  return 1
+}
+
+check_runit() {
+  if [ -z "${IS_RUNIT}" ]; then
+    isrunit
+    IS_RUNIT="$?"
+  fi
+
+  return "${IS_RUNIT}"
+}
+
+install_runit_service() {
+  error "Detected runit, which we do not currently support."
+  exit 3
+}
+
+runit_cmds() {
+  error "Detected runit, which we do not currently support."
+  exit 3
+}
+
+# =====================================================================
+# WSL support functions
+#
+# Cannot be supported, this exists to provide useful error messages.
+
+iswsl() {
+  # If uname -r contains the string WSL, then it's WSL.
+  uname -r | grep -q 'WSL' && return 0
+
+  # If uname -r contains the string Microsoft, then it's WSL.
+  # This probably throws a false positive on CBL-Mariner, but it's part of what MS officially recommends for
+  # detecting if you're running under WSL.
+  uname -r | grep -q "Microsoft" && return 0
+
+  return 1
+}
+
+check_wsl() {
+  if [ -z "${IS_WSL}" ]; then
+    iswsl
+    IS_WSL="$?"
+  fi
+
+  return "${IS_WSL}"
+}
+
+install_wsl_service() {
+  error "We appear to be running in WSL. Netdata cannot be automatically installed as a service under WSL."
+  exit 3
+}
+
+wsl_cmds() {
+  error "We appear to be running in WSL. Netdata cannot be automatically installed as a service under WSL."
+  exit 3
+}
+
+# =====================================================================
+# FreeBSD support functions
+
+enable_freebsd() {
+  if ! sysrc netdata_enable=YES; then
+    warning "Failed to enable netdata service."
+  fi
+}
+
+disable_freebsd() {
+  if ! sysrc netdata_enable=NO; then
+    warning "Failed to disable netdata service."
+  fi
+}
+
+install_freebsd_service() {
+  install_generic_service freebsd "FreeBSD rc.d" /usr/local/etc/rc.d/netdata enable_freebsd disable_freebsd
+}
+
+freebsd_cmds() {
+  NETDATA_START_CMD='service netdata start'
+  NETDATA_STOP_CMD='service netdata stop'
+  NETDATA_INSTALLER_START_CMD='service netdata onestart'
+}
+
+# =====================================================================
+# macOS support functions
+
+install_darwin_service() {
+  info "Installing macOS plist file for launchd."
+  if ! install -C -S -p -m 0644 -o 0 -g 0 system/netdata.plist /Library/LaunchDaemons/com.github.netdata.plist; then
+    error "Failed to copy plist file."
+    exit 4
+  fi
+
+  if ! launchctl load /Library/LaunchDaemons/com.github.netdata.plist; then
+    error "Failed to load plist file."
+    exit 4
+  fi
+}
+
+darwin_cmds() {
+  NETDATA_START_CMD='launchctl start com.github.netdata'
+  NETDATA_STOP_CMD='launchctl stop com.github.netdata'
+}
+
+# =====================================================================
+# Linux support functions
+
+detect_linux_svc_type() {
+  if [ "${SVC_TYPE}" = "detect" ]; then
+    for t in ${LINUX_INIT_TYPES}; do
+      if "check_${t}"; then
+        SVC_TYPE="${t}"
+        break
+      fi
+    done
```

Comment:
For anyone looking at the code who is confused by this construct, it’s iterating over the list of init system types in `${LINUX_INIT_TYPES}`, invoking a callback function for each type by computing the callback name from based on the type name, and then stopping after the first such function that returns true. It can be thought of as functionally equivalent to iterating over a list of function pointers in C, or over a list of function objects in Python and calling each one in turn.

Computed function names are admittedly not the prettiest or easiest to understand construct, but this is at least easier to maintain than the long sequence of conditionals that would be the only other efficient way to do this in shell script.